core/code.py

```diff
@@ -98,77 +99,128 @@ def Call(self, args):
+                    arg_str = arg_str[1:]
 
-        param_name = p.name  # may get hidden __
-        if i < num_args:
-            arg_str = argv[i]
+                val = value.Str(arg_str)
 
-            # If we have myproc(p), and call it with myproc :arg, then bind
-            # __p to 'arg'.  That is, the param has a prefix ADDED, and the arg
-            # has a prefix REMOVED.
-            #
-            # This helps eliminate "nameref cycles".
             if is_out_param:
-                param_name = '__' + param_name
-
-                if not arg_str.startswith(':'):
-                    # TODO: Point to the exact argument.  We got argv but not
-                    # locations.
-                    e_die('Ref param %r expected arg starting with colon : but got %r' %
-                          (p.name, arg_str))
-
-                arg_str = arg_str[1:]
-
-            val = value.Str(arg_str)  # type: value_t
-            #log('%s -> %s', param_name, val)
-        else:
-            # default args were evaluated on definition
-
-            val = proc.defaults[i]
-            if val is None:
-                e_die("No value provided for param %r" % p.name, p.blame_tok)
-
-        if is_out_param:
-            flags = state.SetNameref
-        else:
-            flags = 0
-
-        #log('flags %s', flags)
-        mem.SetValue(lvalue.Named(param_name, p.blame_tok),
-                     val,
-                     scope_e.LocalOnly,
-                     flags=flags)
-
-    num_params = len(sig.word_params)
-    if sig.rest_of_words:
-        r = sig.rest_of_words
-        lval = lvalue.Named(r.name, r.blame_tok)
-
-        items = [value.Str(s) for s in argv[num_params:]]  # type: List[value_t]
-        rest_val = value.List(items)
-        mem.SetValue(lval, rest_val, scope_e.LocalOnly)
-    else:
-        if num_args > num_params:
-            # TODO: Raise an exception?
-            errfmt.Print_(
-                "proc %r expected %d arguments, but got %d" %
-                (proc.name, num_params, num_args), arg0_loc)
-            # This should be status 2 because it's like a usage error.
-            return 2
+                flags = state.SetNameref
+            else:
+                flags = 0
+
+            mem.SetValue(lvalue.Named(param_name, p.blame_tok),
+                         val,
+                         scope_e.LocalOnly,
+                         flags=flags)
+
+        if sig.rest_of_words:
+            rw = sig.rest_of_words
+            items = [value.Str(x)
+                     for x in t.RestWords()]  # type: List[value_t]
+            val = value.List(items)
+
+            mem.SetValue(lvalue.Named(rw.name, rw.blame_tok), val,
+                         scope_e.LocalOnly)
+
+        npos = t.NumPos()
+        for i, p in enumerate(sig.pos_params):
+            if i >= npos and p.default_val:
+                # TODO: can we cache this? would require dependency checking...
```

Comment:
Ah the defaults for word_params are computed in `osh/cmd_eval.py` when the `proc` is **defined**

I think we should keep that?  That's why I mentioned that issue about mutable default args ...

I can help with this ...